Python/sysmodule.c

```diff
@@ -111,6 +112,266 @@ PySys_SetObject(const char *name, PyObject *v)
     }
 }
 
+int
+PySys_Audit(const char *event, const char *argFormat, ...)
+{
+    PyObject *eventName = NULL;
+    PyObject *eventArgs = NULL;
+    PyObject *hooks = NULL;
+    PyObject *hook = NULL;
+    int res = -1;
+    int dtrace = PyDTrace_AUDIT_ENABLED();
+
+    /* N format is inappropriate, because you do not know
+       whether the reference is consumed by the call.
+       Assert rather than exception for perf reasons */
+    assert(!argFormat || !strchr(argFormat, 'N'));
+
+    /* Early exit when no hooks are registered */
+    _Py_AuditHookEntry *e = _PyRuntime.audit_hook_head;
+    PyThreadState *ts = _PyThreadState_GET();
+    PyInterpreterState *is = ts ? ts->interp : NULL;
```

Comment:
How does the rest of the function work without a valid interpreter state? The function is operating on Python objects and at least calls ``PyTyple_New(0)``.

This function should verify that it's called with a valid interpreter state and the GIL held for the current interpreter. Otherwise it's calling for trouble and memory corruption.

Reply:
It used to do that, then when the condition below was rewritten to deal with changes to pystate I updated it incorrectly (this is why I want to get this merged - there are so many changes that cause merge conflicts going on right now...)